python/dgl/nn/pytorch/network_emb.py

```diff
@@ -0,0 +1,220 @@
+    sparse  :int , optional
+        If set to True, gradients w.r.t. to the weight matrix will be sparse. (default: True)
+    """
+
+    def __init__(self,
+                 g,
+                 emb_dimension,
+                 metapath,
+                 window_size,
+                 min_count=0,
+                 negative_samples_size=5,
+                 node_repeat=1,
+                 nid2word=None,
+                 sparse=True
+                 ):
+        super(Metapath2vec, self).__init__()
+        self.g = g
+        self.emb_dimension = emb_dimension
+        self.metapath = metapath
+        self.min_count = min_count
+        self.window_size = window_size
+        self.negative_samples_size = negative_samples_size
+        self.nid2word = nid2word
+        self.sparse = sparse
+        self.node_repeat = node_repeat
+        self.token_count = 0
+        self.walk_dataset = []
+        self.negatives = []
+        self.discards = dict()
+        self.negpos = 0
+        self.NEGATIVE_TABLE_SIZE = 1e8
+        self._extract()
+        self._initTableNegatives()
+
+    def _extract(self):
+
+        assert len(self.metapath) + 1 >= self.window_size
+        word_frequency = dict()
+        word_frequency_filtered = dict()
+
+        ##build whole vocab for all nodes
+        self.token_count = 0
+        self.type2id = dict()
+        self.id2word = dict()
+        wid = 0
+        if self.nid2word is None:
+            for nodetype in self.g.ntypes:
+                self.type2id[nodetype] = dict()
+                for idx in range(self.g.num_nodes(nodetype)):
+                    self.type2id[nodetype][idx] = wid
+                    wid += 1
+        else:
+            for nodetype in self.g.ntypes:
+                self.type2id[nodetype] = dict()
+                for idx in range(self.g.num_nodes(nodetype)):
+                    self.type2id[nodetype][idx] = wid
+                    self.id2word[wid] = self.nid2word[nodetype][idx]
+                    wid += 1
+
+        ## convert the edge metapath to the node metapath
+        edge2des = {}
+        edge2src = {}
+        for comb in self.g.canonical_etypes:
+            edge2des[comb[1]] = comb[2]
+            edge2src[comb[1]] = comb[0]
+        nodespath = []
+        nodespath.append(edge2src[self.metapath[0]])
+        for edge in self.metapath:
+            nodespath.append(edge2des[edge])
+
+        ##start random walk
+        for idx in tqdm.trange(self.g.number_of_nodes(nodespath[0])):
+            traces = random_walk(g=self.g, nodes=[idx, ] * self.node_repeat, metapath=self.metapath)
+            for tr in traces[0].cpu().numpy():
+                line = [self.type2id[nodespath[i]][tr[i]] for i in range(len(tr))]
+                self.walk_dataset.append(line)
+                if len(line) > 1:
+                    for word in line:
+                        self.token_count += 1
+                        word_frequency[word] = word_frequency.get(word, 0) + 1
```

Comment:
You can initialize `word_frequency` with `defaultdict(int)`, which initializes any new `word` with `word_frequency[word] = 0`.